Add support for ECDSA

jderusse opened the pull request:
Fixes #23 

This PR contains 1 BC break `DataSigner::signData` now takes a string alg (RS256, ES384, ...) instead of a int (OPENSSL_ALGO_SHA256). bacause the the method `openssl_sign` returns a format PKCS, wereas EC256 expect a specific format.

tgalopin commented:
Unfortunately, the different BC breaks would mean we would need to release a 2.0 for this feature, which seems a shame given how recent 1.0 is :) (but it could make sense to bump to PHP 7.1 too though). Do you think we could add small BC layers to avoid upgrade issues?

jderusse commented:
Yep.. you're right @tgalopin gonna work on it

Ayesh commented:
Wow this is incredible work @jderusse!

<Bike-shedding>
I suppose the key type (`rsa|ec`) could be deduced from the key strength (2048 bit, etc). >=1024 assumes an RSA key, and 256, 385 (and perhaps `ed` for ed25519) "key sizes"  automatically making it an EC key.

For BC, we can make this CLI option not register for PHP <= 7.1 although I don't think we should give it any love after this year. 

If it sounds like a good idea, I would love to attempt this. 

</Bike-shedding>

tgalopin commented on README.md:
````diff
@@ -85,6 +85,7 @@ config file like
 
 ```yaml
 contact_email: contact@company
+key_type: ec
````
Should we default on `rsa` or `ec`? I think `rsa` is the most popular, perhaps could we put here `key_type: rsa # use "ec" to use a ECDSA key`

jderusse replied:
indeed, example should have been the common use case

tgalopin commented on src/Cli/Command/RunCommand.php:
```diff
@@ -72,7 +75,18 @@ protected function execute(InputInterface $input, OutputInterface $output)
         $config = $this->getConfig(Path::makeAbsolute($input->getArgument('config'), getcwd()));
 
-        $this->register($config['contact_email']);
+        switch ($config['key_type']) {
```
This logic is duplicated 3 times, could it be a good idea to put it in a trait? `KeyOptionCommandTrait`?

jderusse replied:
done

tgalopin commented on src/Cli/Configuration/DomainConfiguration.php:
```diff
@@ -48,7 +48,23 @@ public function getConfigTreeBuilder()
                             return !filter_var($item, FILTER_VALIDATE_EMAIL);
                         })
-                        ->thenInvalid('The email "%s" is not valid.')
+                        ->thenInvalid('The email %s is not valid.')
```
I would keep the quotes, it avoids spacing problems that are difficult to read and debug.

jderusse replied:
The value injected in the placeholder already contain quote. (see https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Config/Definition/Builder/ExprBuilder.php#L189) 

This pr replace
```diff
- The email ""foo"" is not valid.
+ The email "foo" is not valid.
```

tgalopin commented on src/Cli/Configuration/DomainConfiguration.php:
```diff
@@ -48,7 +48,23 @@ public function getConfigTreeBuilder()
+                ->end()
+                ->scalarNode('key_type')
+                    ->info('Type of private key (RSA or EC).')
+                    ->defaultValue('RSA')
```
I guess the default should be in lowercase `rsa`?

jderusse replied:
normalized to lower case. but maybe I should normalized everything to uppercase...

tgalopin commented on src/Cli/Configuration/DomainConfiguration.php:
```diff
@@ -48,7 +48,23 @@ public function getConfigTreeBuilder()
+                        ->ifTrue(function ($item) {
+                            return !\in_array($item, ['rsa', 'ec']);
+                        })
+                        ->thenInvalid('The keyType %s is not valid.')
```
I think it could be useful to add the list of available types in the message:

```
The key type "%s" is not valid. Valid types are: rsa, ec.
```

jderusse replied:
done

tgalopin commented on src/Core/Http/SecureHttpClient.php:
```diff
@@ -136,17 +138,77 @@ public function signedRequest($method, $endpoint, array $payload = [], $returnJs
+                    case 521:
+                        return 'ES512';
+                }
+                throw new AcmeCoreClientException('Private key type is not supported');
```
I think if you don't put any `break`, default will be executed, so this line is not needed IMO.

jderusse replied:
done, even when I don't really like case without break

tgalopin commented on README.md:
````diff
@@ -85,6 +85,7 @@ config file like
 
 ```yaml
 contact_email: contact@company
+key_type: RSA                                          # RSA or EC (for ECDSA). Default "rsa"
````
Default "RSA" perhaps? :stuck_out_tongue:

tgalopin commented on src/Cli/Command/Helper/KeyOptionCommandTrait.php:
```diff
@@ -0,0 +1,33 @@
+            case 'EC':
+                return new EcKeyOption();
+            default:
+                throw new \InvalidArgumentException('the given type of key type is not supported.');
```
This message is kinda weird IMO. May be the message as in config `The keyType "%s" is not valid. Supported types are: RSA, EC`?

jderusse replied:
done

tgalopin commented on src/Core/Http/SecureHttpClient.php:
```diff
@@ -136,17 +138,77 @@ public function signedRequest($method, $endpoint, array $payload = [], $returnJs
+        switch ($parsedKey->getType()) {
+            case OPENSSL_KEYTYPE_RSA:
+                return [
+                    // this order matter
```
I know it's not your code, but this is a good moment to fix this typo IMO :) : `this order matters`.

tgalopin commented on src/Core/Http/SecureHttpClient.php:
```diff
@@ -136,17 +138,77 @@ public function signedRequest($method, $endpoint, array $payload = [], $returnJs
+                ];
+            case OPENSSL_KEYTYPE_EC:
+                return [
+                    // this order matter
```
Same here

jderusse replied:
fixed

tgalopin commented on src/Ssl/Generator/KeyOption.php:
```diff
@@ -0,0 +1,16 @@
+
+namespace AcmePhp\Ssl\Generator;
+
+abstract class KeyOption
```
Shouldn't this be an interface instead? Not sure why an empty abstract class can be useful.

tgalopin commented on src/Ssl/Generator/PrivateKeyGeneratorInterface.php:
```diff
@@ -0,0 +1,43 @@
+    public function generatePrivateKey(KeyOption $keyOption);
+
+    /**
+     * Returns weither or not the instance is able to generator a private key from the given option.
```
`whether` and you can remove the `or not`

tgalopin commented on src/Ssl/Signer/DataSigner.php:
```diff
@@ -21,21 +22,118 @@
+        }
+    }
+
+    private function ECDSAtoDER($signature, $partLength)
```
This method is ... tough :D . Would you have a link to add as a comment to understand the details there? Could be useful to maintain it if needed :) .

jderusse replied:
add pointer to the source in the docblock ;-)

tgalopin commented on tests/Ssl/Generator/KeyPairGeneratorTest.php:
```diff
@@ -28,12 +30,35 @@ public function setUp()
     public function test generateKeyPair generate random instance of KeyPair()
     {
-        $result = $this->service->generateKeyPair(1024);
+        $result = $this->service->generateKeyPair(new RsaKeyOption(1024));
```
I think it would be great to keep a test using the deprecated way to ensure it still works.

jderusse replied:
done

tgalopin commented on src/Ssl/Generator/KeyPairGenerator.php:
```diff
@@ -24,58 +26,47 @@
-                )
-            );
+        if (\is_int($keyOption)) {
+            @trigger_error('Passing a keySize to "generateKeyPair" is deprecated since version 1.0 and will be removed in 2.0. Pass an instance of KeyOption instead', E_USER_DEPRECATED);
```
Could you put 1.1 there instead (and same for other places)? I will publish 1.1 with this in it :) .

tgalopin commented on src/Ssl/Generator/PrivateKeyGeneratorInterface.php:
```diff
@@ -0,0 +1,43 @@
+     *
+     * @return bool
+     */
+    public function supportsPrivateKey(KeyOption $keyOption);
```
Shouldn't this be named `supportsOptions`? Not sure, it's okay if you prefer to keep it this way :) .

jderusse commented:
Hello  @cmorgenstern @rokclimb15 @Ayesh could you please have a look to that PR.
It will be great if you could test it locally and check the generated certificate.

jderusse commented:
@tgalopin I added DSA and DH key generators (not compatible with letsencrypt, but should be part of SSL IMHO).
I also refactored the generators to share logic with a trait. Could you have a 2nd look please (see my last commits or files located in `src/Ssl`. Thanks

tgalopin commented on src/Core/Http/SecureHttpClient.php:
```diff
@@ -136,17 +138,77 @@ public function signedRequest($method, $endpoint, array $payload = [], $returnJs
+                    case 256:
+                    case 384:
+                        return 'ES'.$parsedKey->getBits();
+                    case 521:
```
512?

jderusse replied:
nop... https://tools.ietf.org/html/rfc4754#section-4 probably a math logic behind this

tgalopin replied:
Okay :) !

tgalopin commented on src/Ssl/Signer/DataSigner.php:
```diff
@@ -21,21 +22,142 @@
+    }
+
+    /**
+     * Convert a ECDSA signature into DER.
```
This method does not seem to be actually used, should we still keep it?

jderusse replied:
removed

rokclimb15 commented on src/Ssl/Generator/EcKey/EcKeyGenerator.php:
```diff
@@ -0,0 +1,47 @@
+
+        return $this->generatePrivateKeyFromOpensslOptions(
+            [
+                'private_key_type' => OPENSSL_KEYTYPE_EC,
```
You might consider adding a `defined('OPENSSL_KEYTYPE_EC')` here and throw an exception if it is `false`. Not all OpenSSL versions support this key type.

jderusse replied:
Documentation says http://php.net/manual/en/openssl.key-types.php
> This constant was added in 5.2.0.

are you sure that some version does expose the constant?

jderusse replied:
nvm. Imissed the note
> This constant is only available when PHP is compiled with OpenSSL 0.9.8+.

jderusse replied:
could you please have a look to my last commit? maybe @tgalopin  too

jderusse commented on src/Ssl/const.php:
```diff
@@ -0,0 +1,3 @@
+<?php
+
+defined('OPENSSL_KEYTYPE_EC') or define('OPENSSL_KEYTYPE_EC', 3);
```
Indeed, if the constant is not defined, then I define it by myself. 

It won't fix by magic the generation of ECDSA (the user have to install the proper extension), but it'll avoid triggering errors for user using RSA keys here https://github.com/acmephp/acmephp/blob/master/src/Ssl/ParsedKey.php#L53

I'll add a check to throw early a proper exception if the user don't have the right version of the extension

jderusse replied:
Replaced the const file by a constraint on the openssl lib version.
Added a test on the version of PHP

tgalopin commented:
I'll have a look tomorrow :)

jderusse commented:
Thank you for the review @tgalopin @rokclimb15
